canonical/grafana-agent-k8s-operator
# Fix label treatment of related charms

## Issue
This charm overwrites existing topology labels with its own topology labels. This happens because:
1. This charm doesn't have cos-tool, so injection of label matchers is skipped.
2. This charm has an outdated labeling logic: here, all alert labels are being [replaced](https://github.com/canonical/grafana-agent-k8s-operator/blob/5955b98c3ced5f502a03b3ea9948dd4f0f4e1f2a/lib/charms/loki_k8s/v0/loki_push_api.py#L761) by this charm's labels, whereas in cos-lib [only missing labels are added](https://github.com/canonical/cos-lib/blob/486945531c3a95253c2ee3886204071b176e7841/src/cosl/rules.py#L298).
3. This charm has two different use cases for rules that are currently not captured well:
  1. Like most charms, this charm has built-in rules and dashboards that need to be enriched with the charm's topology labels.
  2. Like cos-config, this charm needs to forward 3rd-party rules and dashboards [without changing the labels](https://github.com/canonical/cos-configuration-k8s-operator/issues/44).

## Solution
- [x] Add cos-tool to charm.
- [x] Fetch lib loki_push_api.v1

Fixes #275.


## Context
This PR goes in tandem with:
- https://github.com/canonical/loki-k8s-operator/pull/353


## Testing Instructions
See https://github.com/canonical/loki-k8s-operator/pull/353.
```
Revision 61 of 'grafana-agent-k8s' released to edge/pr-277 (attaching resources: 'agent-image' r38)
```

## Upgrade Notes
No change further action beyond `juju refresh` is needed – labels should be corrected after an upgrade.
